File: frontend/src/api/hitl.ts
/**
 * HITL (Human-in-the-Loop) API Client
 * 
 * Provides interface for HITL confirmation flow:
 * - List pending confirmation requests
 * - Submit user confirmation (approve/reject)
 * - Get request details
 */
import apiClient from './client'

// Types
export interface HITLRequest {
  id: string  // alias for request_id
  sessionId: string  // alias for session_id
  type: string  // alias for operation
  title?: string
  description: string
  context: Record<string, any>
  status: 'pending' | 'approved' | 'rejected'
  riskLevel?: 'low' | 'medium' | 'high'
  createdAt: string  // alias for created_at
}

// API response types (snake_case from backend)
interface HITLRequestRaw {
  request_id: string
  session_id: string
  operation: string
  description: string
  context: Record<string, any>
  created_at: string
}

interface CreateHITLPayload {
  type: string
  title?: string
  description: string
  context?: Record<string, any>
  riskLevel?: 'low' | 'medium' | 'high'
}

// Transform raw API response to frontend format
function transformRequest(raw: HITLRequestRaw): HITLRequest {
  return {
    id: raw.request_id,
    sessionId: raw.session_id,
    type: raw.operation,
    title: raw.operation,
    description: raw.description,
    context: raw.context,
    status: 'pending',
    createdAt: raw.created_at,
  }
}

export interface HITLConfirmPayload {
  approved: boolean
  user_feedback?: string | null
}

export interface HITLConfirmResponse {
  request_id: string
  approved: boolean
  user_feedback: string | null
  responded_at: string
}

// API Functions
export const hitlApi = {
  /**
   * Create a new HITL intervention request
   */
  async create(sessionId: string, payload: CreateHITLPayload): Promise<HITLRequest> {
    try {
      const response = await apiClient.post<HITLRequestRaw>(
        `/api/v1/sessions/${sessionId}/hitl`,
        {
          operation: payload.type,
          description: payload.description,
          context: payload.context || {},
        }
      )
      return transformRequest(response.data)
    } catch {
      // Fallback for demo/offline mode
      return {
        id: 'manual-' + Date.now(),
        sessionId,
        type: payload.type,
        title: payload.title,
        description: payload.description,
        context: payload.context || {},
        status: 'pending',
        riskLevel: payload.riskLevel,
        createdAt: new Date().toISOString(),
      }
    }
  },

  /**
   * List all pending HITL requests for a session
   */
  async listPending(sessionId: string): Promise<HITLRequest[]> {
    const response = await apiClient.get<HITLRequestRaw[]>(
      `/api/v1/sessions/${sessionId}/hitl/pending`
    )
    return response.data.map(transformRequest)
  },

  /**
   * Submit confirmation for a HITL request
   */
  async confirm(
    requestId: string,
    payload: HITLConfirmPayload
  ): Promise<HITLConfirmResponse> {
    const response = await apiClient.post<HITLConfirmResponse>(
      `/api/v1/hitl/${requestId}/confirm`,
      payload
    )
    return response.data
  },

  /**
   * Get HITL request details by ID
   */
  async getRequest(requestId: string): Promise<HITLRequest> {
    const response = await apiClient.get<HITLRequestRaw>(
      `/api/v1/hitl/${requestId}`
    )
    return transformRequest(response.data)
  },
}

export default hitlApi

File: backend/app/api/v1/hitl.py
"""
HITL (Human-in-the-Loop) API endpoints.
"""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from redis.asyncio import Redis

from app.core.redis import get_redis
from app.services.hitl_service import HITLService

router = APIRouter()


class HITLConfirmRequest(BaseModel):
    """Request body for HITL confirmation."""

    approved: bool
    user_feedback: str | None = None


class HITLRequestResponse(BaseModel):
    """Response for HITL request."""

    request_id: str
    session_id: str
    operation: str
    description: str
    context: dict
    created_at: str


class HITLConfirmResponse(BaseModel):
    """Response for HITL confirmation."""

    request_id: str
    approved: bool
    user_feedback: str | None = None
    responded_at: str


@router.get("/sessions/{session_id}/hitl/pending")
async def list_pending_hitl_requests(
    session_id: str,
    redis: Redis = Depends(get_redis),
) -> list[HITLRequestResponse]:
    """
    List all pending HITL requests for a session.

    This endpoint allows the UI to poll for pending confirmation requests.
    """
    service = HITLService(redis)
    requests = await service.list_pending_requests(session_id)

    return [
        HITLRequestResponse(
            request_id=req.request_id,
            session_id=req.session_id,
            operation=req.operation,
            description=req.description,
            context=req.context,
            created_at=req.created_at.isoformat(),
        )
        for req in requests
    ]


@router.post("/hitl/{request_id}/confirm")
async def confirm_hitl_request(
    request_id: str,
    body: HITLConfirmRequest,
    redis: Redis = Depends(get_redis),
) -> HITLConfirmResponse:
    """
    Submit user confirmation for a HITL request.

    Args:
        request_id: HITL request ID
        body: Confirmation details

    Returns:
        Confirmation response
    """
    service = HITLService(redis)

    try:
        response = await service.submit_response(
            request_id=request_id,
            approved=body.approved,
            user_feedback=body.user_feedback,
        )

        return HITLConfirmResponse(
            request_id=response.request_id,
            approved=response.approved,
            user_feedback=response.user_feedback,
            responded_at=response.responded_at.isoformat(),
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/hitl/{request_id}")
async def get_hitl_request(
    request_id: str,
    redis: Redis = Depends(get_redis),
) -> HITLRequestResponse:
    """Get HITL request details by ID."""
    service = HITLService(redis)
    request = await service.get_request(request_id)

    if not request:
        raise HTTPException(status_code=404, detail="HITL request not found or expired")

    return HITLRequestResponse(
        request_id=request.request_id,
        session_id=request.session_id,
        operation=request.operation,
        description=request.description,
        context=request.context,
        created_at=request.created_at.isoformat(),
    )
